cms/djangoapps/contentstore/rest_api/v2/views/downstreams.py

```diff
@@ -137,6 +146,79 @@ def get(self, request: _AuthenticatedRequest, course_key_string: str):
         return Response(serializer.data)
 
 
+@view_auth_classes()
+class DownstreamContextListView(DeveloperErrorViewMixin, APIView):
+    """
+    Serves library block->course->container links
+    """
+    def get(self, request: _AuthenticatedRequest, usage_key_string: str) -> Response:
+        """
+        Fetches downstream links for given publishable entity
+        """
+        try:
+            usage_key = UsageKey.from_string(usage_key_string)
+        except InvalidKeyError as exc:
+            raise ValidationError(detail=f"Malformed usage key: {usage_key_string}") from exc
+
+        links = (
+            PublishableEntityLink
+                .get_by_upstream_usage_key(upstream_usage_key=usage_key)
+                .order_by("downstream_context_key", "downstream_parent_usage_key")
+                .values( "downstream_usage_key", "downstream_context_key", "downstream_parent_usage_key")
+        )
+
+        context_key_list = set(link["downstream_context_key"] for link in links)
+
+        courses_display_name = dict(
+            CourseOverview.objects.filter(id__in=context_key_list).values_list('id', 'display_name')
+        )
+
+        result = []
+        for context_key, links_by_context in groupby(links, lambda link: link["downstream_context_key"]):
+            if context_key not in courses_display_name:
+                raise BadDownstream(_("Course {context_key} not found").format(context_key=context_key))
+
+            course_link = {
+                "id": str(context_key),
+                "display_name": courses_display_name[context_key],
+                "url": reverse_course_url("course_handler", context_key),
+                "containers": []
+            }
+
+            for parent_key, links_by_unit  in groupby(links_by_context, lambda link: link["downstream_parent_usage_key"]):
+                try:
+                    parent = modulestore().get_item(parent_key)
```

Comment:
We may have a performance issue here if the component is used in a lot of places. Is there a better way to get this data in batches from a usage_key list?

Reply:
@rpenido How about fetching names from meilisearch index in the frontend instead like we are doing [here](https://github.com/open-craft/frontend-app-course-authoring/blob/9c8b51dcd9a5ddf0ae1f71b4e0f1f417a056d868/src/course-libraries/CourseLibraries.tsx#L146-L152)? We can then simplify this whole view into just sending links directly from db (no need to fetch course names as well).